cpp/src/sampling/rw_traversals.hpp

```diff
@@ -96,6 +104,243 @@ value_t const* raw_const_ptr(device_const_vector_view<value_t>& dv)
+    //
+    void* ptr_d_temp_storage{nullptr};
+    size_t temp_storage_bytes = 0;
+    cub::DeviceSegmentedReduce::Reduce(ptr_d_temp_storage,
+                                       temp_storage_bytes,
+                                       values,
+                                       d_aggregate_weights_.data(),
+                                       num_vertices,
+                                       offsets,
+                                       offsets + 1,
+                                       aggregator_,
+                                       initial_value_,
+                                       handle_.get_stream());
+    // Allocate temporary storage
+    //
+    rmm::device_uvector<std::byte> d_temp_storage(temp_storage_bytes, handle_.get_stream());
+    ptr_d_temp_storage = d_temp_storage.data();
+
+    // Run reduction:
+    //
+    cub::DeviceSegmentedReduce::Reduce(ptr_d_temp_storage,
+                                       temp_storage_bytes,
+                                       values,
+                                       d_aggregate_weights_.data(),
+                                       num_vertices,
+                                       offsets,
+                                       offsets + 1,
+                                       aggregator_,
+                                       initial_value_,
+                                       handle_.get_stream());
+  }
+
+  // no need for type-erasure, as this is only used internally:
+  //
+  visitors::return_t const& get_result(void) const override { return ret_unused_; }
+
+  rmm::device_uvector<weight_t>&& get_aggregated_weights(void)
+  {
+    return std::move(d_aggregate_weights_);
+  }
+
+  rmm::device_uvector<weight_t> const& get_aggregated_weights(void) const
+  {
+    return d_aggregate_weights_;
+  }
+
+ private:
+  raft::handle_t const& handle_;
+  rmm::device_uvector<weight_t> d_aggregate_weights_;
+  binary_op_t aggregator_;
+  weight_t initial_value_{0};
+  visitors::return_t ret_unused_{};  // necessary to silence a concerning warning
+};
+
+// Biased RW selection logic:
+//
+// FIXME:
+// 1. move sum weights calculation into selector;
+// 2. pass graph_view to constructor;
+//
+template <typename graph_type, typename real_t>
+struct biased_selector_t {
+  using vertex_t = typename graph_type::vertex_type;
+  using edge_t   = typename graph_type::edge_type;
+  using weight_t = typename graph_type::weight_type;
+
+  struct sampler_t {
+    sampler_t(edge_t const* ro,
+              vertex_t const* ci,
+              weight_t const* w,
+              weight_t const* ptr_d_sum_weights)
+      : row_offsets_(ro), col_indices_(ci), values_(w), ptr_d_sum_weights_(ptr_d_sum_weights)
+    {
+    }
+
+    // pre-conditions:
+    //
+    // 1. (indices, weights) are assumed to be reshuflled
+    //    so that weights(neighborhood(src_v)) are ordered
+    //    increasingly; too expesnive to check this here;
```

Comment:
`expesnive` should be `expensive`

Is this still a precondition?  I know we discussed how we probably don't need this precondition, but I'm not sure if this implementation still requires that.

Reply:
No, it's not a pre-condition anymore. I'll remove that comment.

Reply:
Done.